### ISSUE TITLE
Implementing Serde traits

### ISSUE THREAD

**Issue by hmuendel** [thejpster/multi-map#8]
I am using a `MultiMap` in my own struct and want to serialize it into json.
Deriving `Serialize` fails, because `MultiMap` is not implementing it.

I was also not smart enough to make the [remote deriving](https://serde.rs/remote-derive.html) work.

Would it make sense to implement it? Should `MultiMap` just derive it or rather provide a custom implementation, where maybe the alt key is part of the object?
Should it be behind a feature toggle to not pull in unneeded dependencies?

**Comment by thejpster** [comment]
Yeah, if it requires dependencies I'd put it behind a "serde" feature flag.

How would you serialise the second key? Would you make the JSON value a dictionary of key2 and the actual value?

`{ "some_key1": { "key2": "some_key2", "value": "some value" } }`

**Comment by hmuendel** [comment]
That was my first  idea, but it might not be general enough and should be somehow configurable. In my use case, I don't care about the secondary key.

I am also not sure how serde derive would serialise the tuple, I will try it.
In an ideal world the user might pick his serialisation style on a type basis maybe?
I could think of only primary key, only secondary key, and then embedded style as you described it above. Or just go with the tuple. however this will look.
Is this possible to make it on a type declaration basis? Maybe with a MultiMap specific macro?

**Comment by hmuendel** [comment]
if I derive the traits like this
```rust
#[derive(Eq, Serialize, Deserialize)]
pub struct MultiMap<K1: Eq + Hash, K2: Eq + Hash, V> {
    #[serde(flatten)]
    value_map: HashMap<K1, (K2, V)>,
    #[serde(skip)]
    key_map: HashMap<K2, K1>,
}
```
The tuple is transformed into an array and the serialised result looks like this:
```{"2":["Two","Zwei"],"3":["Three","Drei"],"1":["One","Eins"]}```
But I guess for the deserialisation we still would need a custom serialiser, to construct the key map 
correctly. I will try somethings

**Comment by thejpster** [comment]
Hmm, I wonder if adding `impl From<HashMap<K1, (K2, V)>>` would help? I don't know if Serde understands than an array of length two can be mapped two a 2-tuple.

**Comment by hmuendel** [comment]
This was a very helpful remark. 
It would work. But I had to add the `Clone` trait bound to the keys. I guess because they live in both maps?
```rust
#[derive(Eq, Serialize, Deserialize)]
#[serde(from = "HashMap<K1, (K2, V)>")]
pub struct MultiMap<K1, K2, V>
where
    K1: Eq + Hash + Clone,
    K2: Eq + Hash + Clone,
{
    #[serde(flatten)]
    value_map: HashMap<K1, (K2, V)>,
    #[serde(skip)]
    key_map: HashMap<K2, K1>,
}

impl<K1, K2, V> From<HashMap<K1, (K2, V)>> for MultiMap<K1, K2, V>
where
    K1: Eq + Hash + Clone,
    K2: Eq + Hash + Clone,
{
    fn from(tuple_map: HashMap<K1, (K2, V)>) -> Self {
        let mut m = MultiMap::with_capacity(tuple_map.len());
        for (k1, (k2, v)) in tuple_map {
            m.insert(k1, k2, v);
        }
        m
    }
}
```
I will put it behind a feature flag and make a PR.

**Comment by hmuendel** [comment]
Ah, the impl block requires `Clone` so I guess adding it to the struct is fine?

**Comment by thejpster** [comment]
> Ah, the impl block requires `Clone` so I guess adding it to the struct is fine?

Yeah, it was probably a mistake that it was missing.

**Comment by thejpster** [comment]
Closed via #11
